pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,394 @@
+	batchSize := uint32(bfs.cfg.BatchSize)
+
+	for {
+		select {
+		case <-bfs.quit:
+			bfs.log.Info("NeoFS block fetcher service: stopping NeoFS data fetching operation.")
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex+batchSize-1), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+
+			blockOids, err := bfs.search(ctx, prm)
+			cancel()
+			bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: found %d blocks from index %d to %d", len(blockOids), startIndex, startIndex+batchSize-1))
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to find block with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: no block found with index %d, stopping.", startIndex))
+				return errors.New("NeoFS block fetcher service: no block found")
+			}
+			err = bfs.fetchAndProcessBlocks(blockOids)
+			if err != nil {
+				return err
+			}
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchAndProcessBlocks(blockOids []oid.ID) error {
+	var wg sync.WaitGroup
+
+	for i, oidBlock := range blockOids {
+		wg.Add(1)
+		go func(oidBlock oid.ID, i int) {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			defer cancel()
+			rc, err := bfs.get(ctx, oidBlock.String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to fetch block %s", oidBlock.String()), zap.Error(err))
+				return
+			}
+
+			err = bfs.processBlock(rc, i)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to process block with OID %s", oidBlock.String()), zap.Error(err))
+			}
+		}(oidBlock, i)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (bfs *Service) fetchDataWithOid() error {
+	startIndex := bfs.chain.BlockHeight() / 2000
+	skip := bfs.chain.BlockHeight() % 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.OidAttribute, fmt.Sprintf("%d", startIndex), object.MatchStringEqual)
+			prm.SetFilters(filters)
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			blockOidsObject, err := bfs.search(ctx, prm)
+			cancel()
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to find 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: no %s object found with index %d, stopping.", bfs.cfg.OidAttribute, startIndex))
```

Comment:
Error logs in this method also may be moved to upper level. Just return an appropriate error from the method itself.